src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,546 @@
+      auto this_value = Get(resource_id);
+      auto other_value = other.Get(resource_id);
+      if (this_value > other_value) {
+        Set(resource_id, other_value);
+      }
+    }
+  }
+
+  /// Remove the negative values in this set.
+  void RemoveNegative() {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i] < 0) {
+        predefined_resources_[i] = 0;
+      }
+    }
+    for (auto it = custom_resources_.begin(); it != custom_resources_.end();) {
+      if (it->second < 0) {
+        custom_resources_.erase(it++);
+      } else {
+        it++;
+      }
+    }
+  }
+
+  /// Return the number of resources in this set.
+  size_t Size() const {
+    size_t size = custom_resources_.size();
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      if (predefined_resources_[i] != 0) {
+        size++;
+      }
+    }
+    return size;
+  }
+
+  /// Return true if this set is empty.
+  bool IsEmpty() const { return Size() == 0; }
+
+  /// Return a set that contains all resource ids in this set.
+  absl::flat_hash_set<ResourceID> ResourceIds() const {
+    absl::flat_hash_set<ResourceID> res;
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i] != 0) {
+        res.insert(ResourceID(i));
+      }
+    }
+    for (auto &entry : custom_resources_) {
+      res.insert(ResourceID(entry.first));
+    }
+    return res;
+  }
+
+  /// Return a map from the resource ids to the values.
+  absl::flat_hash_map<ResourceID, FixedPoint> ToMap() const {
+    absl::flat_hash_map<ResourceID, FixedPoint> res;
+    for (auto &resource_id : ResourceIds()) {
+      res.emplace(resource_id, Get(resource_id));
+    }
+    return res;
+  }
+
+  ResourceRequest &operator=(const ResourceRequest &other) {
+    this->predefined_resources_ = other.predefined_resources_;
+    this->custom_resources_ = other.custom_resources_;
+    this->requires_object_store_memory_ = other.requires_object_store_memory_;
+    return *this;
+  }
+
+  ResourceRequest operator+(const ResourceRequest &other) {
+    ResourceRequest res = *this;
+    res += other;
+    return res;
+  }
+
+  ResourceRequest operator-(const ResourceRequest &other) {
+    ResourceRequest res = *this;
+    res -= other;
+    return res;
+  }
+
+  ResourceRequest &operator+=(const ResourceRequest &other) {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      predefined_resources_[i] += other.predefined_resources_[i];
+    }
+
+    for (auto &entry : other.custom_resources_) {
+      auto it = custom_resources_.find(entry.first);
+      if (it != custom_resources_.end()) {
+        it->second += entry.second;
+        if (it->second == 0) {
+          custom_resources_.erase(it);
+        }
+      } else {
+        custom_resources_.emplace(entry.first, entry.second);
+      }
+    }
+    return *this;
+  }
+
+  ResourceRequest &operator-=(const ResourceRequest &other) {
```

Comment:
I feel we expect `this` and `other` have the same set of resources? I'm not sure if it's a sensible to do things like `ResourceRequest(A, B, C) -/+ ResourceRequest(D, E, F)`

Reply:
I think it's valid. Because we do dynamically adding new resources (PG resources for example) to node.